rust/src/client.rs

```diff
@@ -0,0 +1,166 @@
+        })
+    }
+
+    /// Start the `Client` loop
+    ///
+    /// Returns `Err(err)` if `ClientError` `err` occurred
+    /// NOTE in the future this may iterate only a fixed number of times, at the
+    /// end of which this returns `Ok(())`
+    pub async fn start(&mut self) -> Result<(), ClientError> {
+        loop {
+            // any error that bubbles up will finish off the client
+            if let Err(e) = self.per_round().await {
+                break Err(e)
+            }
+        }
+    }
+
+    /// Client duties within a round
+    async fn per_round(&mut self) -> Result<(), ClientError> {
+        let round_params: Arc<RoundParameters> = loop {
+            if let Some(round_params) =
+                self.handle.get_round_parameters().await
+            {
+                break round_params
+            }
+            self.interval.tick().await;
+        };
+        let coord_pk = round_params.pk;
+        let round_seed: &[u8] = round_params.seed.as_slice();
+        self.particip
+            .compute_signatures(round_seed);
+        let (sum_frac, upd_frac) = (round_params.sum, round_params.update);
+        match self
+            .particip
+            .check_task(sum_frac, upd_frac)
+        {
+            Task::Sum    =>
+                self.summer(coord_pk).await,
+            Task::Update =>
+                self.updater(coord_pk).await,
+            Task::None   =>
+                self.unselected().await,
+        }
+    }
+
+    /// Duties for unselected clients
+    async fn unselected(&mut self) -> Result<(), ClientError> {
+        // TODO await global model; save it
+        // end of round
+        Ok(())
+    }
+
+    /// Duties for clients selected as summers
+    async fn summer(&mut self, coord_pk: CoordinatorPublicKey)
+                    -> Result<(), ClientError>
+    {
+        let sum1_msg: Vec<u8> = self
+            .particip
+            .compose_sum_message(&coord_pk);
+        self.handle
+            .send_message(sum1_msg)
+            .await;
+
+        let pk = self.particip.get_encr_pk();
+        let seed_dict_ser: Arc<Vec<u8>> = loop {
+            if let Some(seed_dict_ser) = self.handle.get_seed_dict(pk).await {
+                break seed_dict_ser
+            }
+            self.interval.tick().await;
+        };
+        let seed_dict: SeedDict = bincode::deserialize(&seed_dict_ser[..])
+            .map_err(ClientError::DeserialiseErr)?;
+        let sum2_msg: Vec<u8> = self
+            .particip
+            .compose_sum2_message(&coord_pk, &seed_dict)
+            .map_err(ClientError::ParticipantErr)?;
+        self.handle
+            .send_message(sum2_msg)
+            .await;
+
+        // job done, unselect
+        self.unselected()
+            .await
+    }
+
+    /// Duties for clients selected as updaters
+    async fn updater(&mut self, coord_pk: CoordinatorPublicKey)
+                     -> Result<(), ClientError>
+    {
+        // TODO train a model update...
```

Comment:
I think in the updater function the updater must first download the global model before it can train on it. The idea I had in mind was that the identifier for the global model (not sure yet if it will be an URL or just the hash of the global model) will be provided via the round parameters.

Reply:
that could work. but if you made it a round parameter, what would be the very first global model? in the old Python coordinator, we eventually decided that the coordinator should not be the one initialising the global model, IIRC. so if we made the global model a round parameter, the very first round ends up being a special case (which is OK, just not ideal). instead we could make the global model available not at the start of a round, but at the end, so after aggregation.

Reply:
True, that would solve the first round issue we had. I'm just not sure how we then provide the "url" to that global model. The advantage to provide the global model at the start of the round is that we can decide later that only the updater of the current round download the global model.